### PR TITLE
[Testcase Refactoring] Add HardwareClassification labels to test_dims.py

### PR DESCRIPTION
**Pull request by syg1996419**
## Summary

Refactor `test/functorch/test_dims.py` to properly separate CUDA-specific
code from device-agnostic tests, and add `HardwareClassification` labels.

### 1. Extract TestBase for shared code

Extract `TestBase(TestCase)` containing the `attn` helper method, shared
by both `TestMin` and `TestMinCudaOnly`.

### 2. Move CUDA-only code to TestMinCudaOnly

Create `TestMinCudaOnly(TestBase)` with:
- `hw_classification = HardwareClassification.CUDA`
- CUDA-specific `setUp`/`tearDown` (memory tracking)
- `gpu_time` method (moved from module-level function)
- `test_attn_cuda` (moved from TestMin)
- `instantiate_device_type_tests(only_for=("cuda",))`

### 3. Add HardwareClassification labels

- `TestMin(TestBase)`: `GENERIC`
- `TestMinFunctorchOnly(TestMin)`: `GENERIC`
- `TestMinCudaOnly(TestBase)`: `CUDA`

### 4. Clean up

- Remove module-level `gpu_time` function
- Remove `test_attn_cuda` from `skip_functorch_only` list
- Remove unused imports (`unittest`, `skipIf`, `IS_LINUX`, `IS_WINDOWS`,
  `TEST_CUDA`, `TEST_WITH_ROCM`, `TEST_WITH_SLOW`)
- Add `instantiate_device_type_tests` import

## Test

Verified: `python test/functorch/test_dims.py -v` — 66 tests, 3 skipped (expected), all pass.

### CODE REVIEW

File: test/functorch/test_dims.py
Comment (xuyun44) [inline test/functorch/test_dims.py]: This function is cuda only

Comment (xuyun44) [inline test/functorch/test_dims.py]: Here existing cuda only branches determination

Comment (xuyun44) [inline test/functorch/test_dims.py]: Here existing cuda only branches determination

Comment (xuyun44) [inline test/functorch/test_dims.py]: May add cudaOnly class inherit class TestMin.

Comment (xuyun44) [inline test/functorch/test_dims.py]: after move test_attn_cuda to cuda only class, remove "test_attn_cuda" here

**Comment by syg1996419** [comment]
Move CUDA-specific code out of the GENERIC TestMin class into a new
TestMinCudaOnly class:

- Move `gpu_time` from module-level function to TestMinCudaOnly method
- Move `test_attn_cuda` from TestMin to TestMinCudaOnly
- Move CUDA memory tracking (setUp/tearDown) from TestMin to TestMinCudaOnly
- Update `gpu_time` calls in `attn` helper from `gpu_time()` to `self.gpu_time()`
- Add `hw_classification = HardwareClassification.CUDA` to TestMinCudaOnly
- Add `hw_classification = HardwareClassification.GENERIC` to TestMin and TestMinFunctorchOnly
- Remove `test_attn_cuda` from `skip_functorch_only` list

TestMin (GENERIC) no longer has any CUDA hard dependencies. All CUDA
code is isolated in TestMinCudaOnly with class-level skip decorators.

Device-specific exceptions:
- `device=\"cuda\"` in test_attn_cuda: tests BERT attention with large
  matrices on CUDA, guarded by class-level @skipIf(not TEST_CUDA)
- TestMinCudaOnly does not use instantiate_device_type_tests: the test
  is intentionally CUDA-only, not cross-device parameterized

Verified: python test/functorch/test_dims.py -v"

**Comment by xuyun44** [comment]
do you run code check locally? if done, lgtm
1. pip install lintrunner
2. lintrunner init
3. lintrunner -a

**Comment by syg1996419** [comment]
> do you run code check locally? if done, lgtm
> 
> 1. pip install lintrunner
> 2. lintrunner init
> 3. lintrunner -a

yes. I ran lintruuner -a on the srever,then synced the linted file to local. code checks completed

**Review by xuyun44 — APPROVED** [review]

File: test/functorch/test_dims.py [hunk 1]
```diff
@@ -688,10 +640,70 @@ def test_big_split(self):
...
-skip_functorch_only = ["test_time_mm_fuse", "test_attn_cuda"]
+skip_functorch_only = ["test_time_mm_fuse"]
+
+
+@unittest.skipIf(
+    IS_LINUX or TEST_WITH_ROCM or TEST_WITH_SLOW or IS_WINDOWS,
+    "https://github.com/pytorch/pytorch/issues/86710",
+)
+@skipIf(not TEST_CUDA, "no CUDA")
+class TestMinCudaOnly(TestMin):
```
Comment (xuyun44) [inline test/functorch/test_dims.py]: instantiate_device_type_tests(
    TestMinCudaOnly,
    globals(),
    only_for=(
        "cuda",
    ),
)
maybe a better chose instead of skipIf

Comment (syg1996419) [reply test/functorch/test_dims.py]: I chose skipIf over instantiate_device_type_tests because TestMinCudaOnly inherits from TestMin with ~30 test methods lacking device param. Using instantiate_device_type_tests would require duplicating the attn helper (~120 lines) into a standalone class. skipIf + hw_classification = CUDA provides equivalent gating with less duplication.

**Review by xuyun44 — APPROVED** [review]

File: test/functorch/test_dims.py [hunk 2]
```diff
@@ -688,10 +635,190 @@ def test_big_split(self):
...
+        #     wait=0,
+        #     warmup=1,
+        #     active=2), on_trace_ready=tensorboard_trace_handler(name), with_stack=True) as profiler:
+        #     for _ in range(3):
+        #         lmb()
+        #         profiler.step()
+        print(name, elapsed / r)
+        return elapsed / r
+
+    def attn(
```
Comment (xuyun44) [inline test/functorch/test_dims.py]: Let's split a public class, put non test headers such as 'def attn' into the public class, and then TestMin and TestMinCudaOnly inherit the public class respectively. 
This can solve the problem copy attn into the new class

File: test/functorch/test_dims.py [hunk 3]
```diff
@@ -688,10 +637,68 @@ def test_big_split(self):
...
 
-skip_functorch_only = ["test_time_mm_fuse", "test_attn_cuda"]
+skip_functorch_only = ["test_time_mm_fuse"]
+
+
+class TestMinCudaOnly(TestBase):
+    hw_classification = HardwareClassification.CUDA
+
+    def setUp(self):
+        super().setUp()
```
Comment (xuyun44) [inline test/functorch/test_dims.py:647]: there exist a problem, super().setUp() should from TestMin not from base.
move setUp from TestMin to TestBase, or find a way to slove this problem.
other functions have simlar problem, e.g., tearDown

**Review by xuyun44 — APPROVED** [review]

**Comment by can-gaa-hou** [comment]
@claude review this pr with pr review skill

File: test/functorch/test_dims.py [hunk 4]
```diff
@@ -688,10 +637,68 @@ def test_big_split(self):
...
+        elapsed = b.elapsed_time(e)
+        # with torch.profiler.profile(schedule=torch.profiler.schedule(
+        #     wait=0,
+        #     warmup=1,
+        #     active=2), on_trace_ready=tensorboard_trace_handler(name), with_stack=True) as profiler:
+        #     for _ in range(3):
+        #         lmb()
+        #         profiler.step()
+        print(name, elapsed / r)
+        return elapsed / r
```
Comment (fffrog) [inline test/functorch/test_dims.py]: Hi, please move this function into the basic class via the accelerator-agnostic form: torch.Event

**Review by fffrog — CHANGES_REQUESTED** [review]

File: test/functorch/test_dims.py [hunk 5]
```diff
@@ -688,10 +659,46 @@ def test_big_split(self):
         x.split(l, 0)
 
 
-skip_functorch_only = ["test_time_mm_fuse", "test_attn_cuda"]
+skip_functorch_only = ["test_time_mm_fuse"]
+
+
+class TestMinCudaOnly(TestBase):
```
Comment (fffrog) [inline test/functorch/test_dims.py]:
```suggestion
class TestMinDevice(TestBase):
```

File: test/functorch/test_dims.py [hunk 6]
```diff
@@ -688,10 +659,46 @@ def test_big_split(self):
...
+    hw_classification = HardwareClassification.CUDA
```
Comment (fffrog) [inline test/functorch/test_dims.py]:
```suggestion
    hw_classification = HardwareClassification.ACCELERATOR
```

File: test/functorch/test_dims.py [hunk 7]
```diff
@@ -688,10 +659,46 @@ def test_big_split(self):
...
+
+    def setUp(self):
+        super().setUp()
+        self.mem_allocated = torch.cuda.memory_allocated()
```
Comment (fffrog) [inline test/functorch/test_dims.py]: Using torch.accelerator.memory_allocated() insteadly

File: test/functorch/test_dims.py [hunk 8]
```diff
@@ -688,10 +659,46 @@ def test_big_split(self):
...
+
+    def tearDown(self):
+        extra_memory = torch.cuda.memory_allocated() - self.mem_allocated
```
Comment (fffrog) [inline test/functorch/test_dims.py]: ditto

File: test/functorch/test_dims.py [hunk 9]
```diff
@@ -688,10 +659,46 @@ def test_big_split(self):
...
+        if extra_memory != 0:
+            gc.collect()
+        self.assertEqual(
+            extra_memory,
+            0,
+            lambda msg: f"{msg}\nextra cuda memory left allocated: {extra_memory}",
```
Comment (fffrog) [inline test/functorch/test_dims.py]:
```suggestion
            lambda msg: f"{msg}\nextra accelerator memory left allocated: {extra_memory}",
```

File: test/functorch/test_dims.py [hunk 10]
```diff
@@ -688,10 +659,46 @@ def test_big_split(self):
...
+        )
+        super().tearDown()
+
+    def test_attn_cuda(self, device):
```
Comment (fffrog) [inline test/functorch/test_dims.py]:
```suggestion
    def test_attn(self, device):
```

File: test/functorch/test_dims.py [hunk 11]
```diff
@@ -688,10 +659,46 @@ def test_big_split(self):
...
+            hidden_size=768,
+            sequence_length=128,
+            num_attention_heads=12,
+            device=device,
+            time=measure_perf,
+            linear=torch.nn.Linear,
+        )
+
+
+instantiate_device_type_tests(TestMinCudaOnly, globals(), only_for=("cuda",))
```
Comment (fffrog) [inline test/functorch/test_dims.py]:
```suggestion
instantiate_device_type_tests(TestMinCudaOnly, globals(), except=("cpu"))
```

File: test/functorch/test_dims.py
Comment (fffrog) [inline test/functorch/test_dims.py:705]: adding a new method named `test_time_mm_fuse` here

Comment (fffrog) [inline test/functorch/test_dims.py]: Remove those lines please

**Review by fffrog — CHANGES_REQUESTED** [review]

File: test/functorch/test_dims.py [hunk 12]
```diff
@@ -700,9 +703,40 @@ def tearDown(self):
         functorch.dim.POINTWISE_OPTIMIZE = True
         super().tearDown()
 
+    def test_time_mm_fuse(self):
```
Comment (fffrog) [inline test/functorch/test_dims.py:712]: Hi @syg1996419 

We should define a new function called `test_time_mm_fuse`, which has the same name as the function in the derived class `TestMin`, and we need to add something like `skip` to the new function so that it is skipped at runtime.

File: test/functorch/test_dims.py [hunk 13]
```diff
@@ -688,10 +658,43 @@ def test_big_split(self):
...
+    def tearDown(self):
+        extra_memory = torch.accelerator.memory_allocated() - self.mem_allocated
+        if extra_memory != 0:
+            gc.collect()
+        self.assertEqual(
+            extra_memory,
+            0,
+            lambda msg: f"{msg}\nextra accelerator memory left allocated: {extra_memory}",
+        )
+        super().tearDown()
```
Comment (fffrog) [inline test/functorch/test_dims.py]: Thank you for separating the accelerator-related parts from the derived class.

However, implementing this functionality becomes very difficult after the separation, right?

```Python
        if extra_memory != 0 or len(interesting) != 0:
            import refcycle

            refcycle.garbage().export_image("garbage.pdf")
```

**Review by fffrog — CHANGES_REQUESTED** [review]

File: test/functorch/test_dims.py [hunk 14]
```diff
@@ -688,10 +658,43 @@ def test_big_split(self):
...
+
+    def test_attn(self, device):
```
Comment (fffrog) [inline test/functorch/test_dims.py]:
```suggestion
    def test_attn_device(self, device):
```

File: test/functorch/test_dims.py [hunk 15]
```diff
@@ -105,22 +74,13 @@ def tearDown(self):
             ):
                 interesting.append(o)
 
-        extra_memory = 0
-        if "cuda" in self._testMethodName:
```
Comment (fffrog) [inline test/functorch/test_dims.py:109]:
```Python
if "test_attn_device" in self._testMethodName:
            extra_memory += torch.accelerator.memory_allocated() - self.mem_allocated
```

File: test/functorch/test_dims.py [hunk 16]
```diff
@@ -105,22 +74,13 @@ def tearDown(self):
...
+        if len(interesting) != 0:
             import refcycle
 
             refcycle.garbage().export_image("garbage.pdf")
         gc.collect()
         # assert nolevels, f"cleanup failed? {_n_levels_in_use()}"
-        self.assertEqual(
-            extra_memory,
-            0,
-            lambda msg: f"{msg}\nextra cuda memory left allocated: {extra_memory}",
```
Comment (fffrog) [inline test/functorch/test_dims.py:122]: lambda msg: f"{msg}\nextra accelerator memory left allocated: {extra_memory}",

File: test/functorch/test_dims.py [hunk 17]
```diff
@@ -688,10 +658,43 @@ def test_big_split(self):
         x.split(l, 0)
 
 
-skip_functorch_only = ["test_time_mm_fuse", "test_attn_cuda"]
+class TestMinDevice(TestBase):
+    hw_classification = HardwareClassification.ACCELERATOR
+
+    def setUp(self):
```
Comment (fffrog) [inline test/functorch/test_dims.py]: delete

File: test/functorch/test_dims.py [hunk 18]
```diff
@@ -688,10 +658,43 @@ def test_big_split(self):
...
+        super().setUp()
+        self.mem_allocated = torch.accelerator.memory_allocated()
+
+    def tearDown(self):
```
Comment (fffrog) [inline test/functorch/test_dims.py]: delete

File: test/functorch/test_dims.py [hunk 19]
```diff
@@ -93,8 +64,6 @@ def setUp(self):
         for o in gc.get_objects():
             if isinstance(o, (torch.Tensor, Dim, Tensor, DimList)):
                 self.interesting.add(id(o))
-        if "cuda" in self._testMethodName:
-            self.mem_allocated = torch.cuda.memory_allocated()
```
Comment (fffrog) [inline test/functorch/test_dims.py:97]:
```Python
if "test_attn_device" in self._testMethodName:
    self.mem_allocated = torch.accelerator.memory_allocated()
```

**Review by fffrog — CHANGES_REQUESTED** [review]

**Review by fffrog — APPROVED** [review]
SGTM

**Comment by fffrog** [comment]
@syg1996419 Please wait the ci green, thank you

**Review by fffrog — APPROVED** [review]
